docs/development/etcdoperatortask-design.md

```diff
@@ -0,0 +1,481 @@
+            return result
+        }
+    }
+
+    return ctrlutils.ReconcileAfter(task.Spec.TTLSecondsAfterFinished, "Task completed, waiting for TTL to expire")
+}
+```
+
+
+### Deletion Flow
+
+
+```go
+func (r *Reconciler) triggerTaskDeletionFlow(
+    ctx context.Context,
+    logger logr.Logger,
+    taskObjKey client.ObjectKey,
+    taskHandler tasks.TaskHandler,
+) ctrlutils.ReconcileStepResult {
+    if task.IsCompleted() && !task.IsMarkedForDeletion() {
+        if !task.TTLHasExpired() {
+            return ctrlutils.ReconcileAfter(task.Spec.TTLSecondsAfterFinished, "Task completed, waiting for TTL to expire")
+        }
+    }
+
+    deletionStepFns := []reconcileFn{
+        r.cleanupTaskResources, // Cleanup of any kind of resources used to run the task. Runs the interface method.
+        r.removeTaskFinalizer,
+        r.removeTask, // Removes the CR.
+    }
+    for _, fn := range deletionStepFns {
+        result := fn(ctx, taskObjKey, taskHandler)
+        if ctrlutils.ShortCircuitReconcileFlow(result) {
+            return r.recordTaskIncompleteDeletionOperation(ctx, logger, taskObjKey, result)
+        }
+    }
+    return ctrlutils.DoNotRequeue()
+}
+
+```
+
+## Example: TaskHandler Implementation
+
+
+```go
+// OnDemandSnapshot implements the TaskHandler interface for on-demand snapshots.
+type OnDemandSnapshot struct {
+    client        client.Client
+    httpClient    *http.Client
+    logger        logr.Logger
+    name          string
+    etcdReference types.NamespacedName
+    config        *Config
+}
+// the config provided in the spec will be parsed as follows:
+const ERR_PRECONDITION_CHECK_ON_DEMAND_SNAPSHOT druidv1alpha1.ErrorCode = "ERR_PRECONDITION_CHECK_ON_DEMAND_SNAPSHOT"
+
+
+func New(k8sclient client.Client, logger logr.Logger, task *v1alpha1.EtcdOperatorTask) (operatortask.OperatorTask, error) {
+     
+    config,err := decodeOnDemandSnapshotConfig(task.Spec.Config)
+    if err != nil {
+        return nil, druiderr.WrapError(err, operatortask.ERR_INVALID_CONFIG, operationNew, "failed to decode config for OnDemandSnapshot")
+    }
+    
+    return &OnDemandSnapshot{
+        client:        k8sclient,
+        logger:        logger,
+        name:          task.Name,
+        httpClient:    &http.Client{Timeout: config.Timeout},
+        etcdReference: types.NamespacedName(*task.Spec.EtcdRef),
+        config:        config,
+    }, nil
+}
+
+func (o *OnDemandSnapshot) Run(ctx context.Context) *operatortask.TaskResult {
+    // Implement the snapshot logic
+    etcd := &v1alpha1.Etcd{}
+    if err := o.client.Get(ctx, o.etcdReference, etcd); err != nil {
+        return &operatortask.TaskResult{Description: "Failed to get Etcd resource", Error: err}
+    }
+
+    snapType := "full"
+    if o.config.SnapshotType != nil && *o.config.SnapshotType != "" {
+        snapType = *o.config.SnapshotType
+    }
+
+    o.logger.Info("Snapshot config", "snapshotType", snapType, "timeout", o.config.Timeout)
```

Comment:
At what level will the logs be visible for operator task, Druid Plutono dashbord in seed or Shoot Plutono dashboard?